### PR TITLE
ci: run unit tests on PR

### PR DESCRIPTION
**Pull request by andnp**
Adds a simple github actions workflow to run the testing infrastructure whenever a PR is opened against `main`.

Tests on macos and ubuntu with python versions 3.10 - 3.12

### CODE REVIEW

**Comment by andnp** [comment]
@golnazmes - here is the config for running tests in the CI build. I'll merge this after we've merged your PR

**Review by andy-amii — APPROVED** [review]

**Pull request merged**
